staking-contract/contracts/StakingRewardSystem.sol

```diff
@@ -0,0 +1,221 @@
+            accumulatedPenalties += penalty;
+        }
+        userStake.amount -= _amount;
+        stakingToken.transfer(msg.sender, _amount);
+        emit Withdrawn(msg.sender, _amount, penalty, _rewardToken);
+    }
+
+  
+    // Function to claim rewards based on staked tokens
+    function claimReward(address _rewardToken) external nonReentrant whenNotPaused {
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (!isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        Stake storage userStake = userStakes[msg.sender][_rewardToken];
+        if (userStake.amount == 0) revert NoStakeFound();
+        if (block.timestamp <= userStake.startTime + minStakingPeriod) revert CannotClaimRewardYet();
+
+        uint256 rewardAmount = _calculateReward(userStake);
+        IERC20(_rewardToken).safeTransfer(msg.sender, rewardAmount);
+        userStake.startTime = block.timestamp;
+        emit RewardClaimed(msg.sender, rewardAmount, _rewardToken);
+    }
+
+    // Internal function to calculate the reward based on staking duration
+    function _calculateReward(Stake storage userStake) internal view returns (uint256) {
+        uint256 stakingDuration = block.timestamp - userStake.startTime;
+        return (userStake.amount * rewardRate * stakingDuration) / 1e18;
+    }
+
+    // Function for the contract owner to withdraw accumulated penalties
+    function withdrawPenalties() external onlyOwner nonReentrant whenNotPaused {
+        uint256 amountToWithdraw = accumulatedPenalties;
+        if (amountToWithdraw == 0) revert NoPenaltiesToWithdraw();
+
+        accumulatedPenalties = 0;
+        stakingToken.transfer(owner(), amountToWithdraw);
+        emit PenaltiesWithdrawn(owner(), amountToWithdraw);
+    }
+
+    // Function to add a new reward token
+    function addRewardToken(address _rewardToken) external onlyOwner {
+        _addRewardToken(_rewardToken);
+    }
+
+    // Internal function to add a reward token and mark it as valid
+    function _addRewardToken(address _rewardToken) internal {
+        if (_rewardToken == address(0)) revert InvalidInput("Reward token address cannot be zero");
+        if (isRewardToken[_rewardToken]) revert InvalidRewardToken();
+        rewardTokens.push(_rewardToken);
+        isRewardToken[_rewardToken] = true;
+        emit RewardTokenAdded(_rewardToken, msg.sender);
+    }
+
+    // Function to set the minimum staking period
+    function setMinStakingPeriod(uint256 _period) external onlyOwner {
+        if (_period <= 0) revert InvalidInput("Minimum staking period must be greater than zero");
+        uint256 prevminStakingPeriod=minStakingPeriod;
+        minStakingPeriod = _period*86400;
+        emit MinStakingPeriodUpdated(prevminStakingPeriod,minStakingPeriod, msg.sender);
+    }
+
+    // Function to set the early withdrawal penalty percentage
+    function setEarlyWithdrawalPenalty(uint256 _penalty) external onlyOwner {
```

Comment:
This comment is still not attended - https://github.com/gajendra-0180/Onboarding-Task/pull/1#discussion_r1738308188